File: exam-system-java/server/src/main/java/com/yyjy/exam/user/service/EmailService.java
package com.yyjy.exam.user.service;

import com.yyjy.exam.common.exception.BusinessException;
import com.yyjy.exam.common.property.MailProperties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;

@Service
public class EmailService {

    private static final Logger log = LoggerFactory.getLogger(EmailService.class);

    private final JavaMailSender mailSender;
    private final MailProperties mailProperties;

    public EmailService(JavaMailSender mailSender, MailProperties mailProperties) {
        this.mailSender = mailSender;
        this.mailProperties = mailProperties;
    }

    public void sendVerificationLink(String to, String uuid) {
        String link = mailProperties.confirmUrl() + "?id=" + uuid;
        String subject = "【考试系统】请验证您的邮箱";
        String content = """
                <div style="font-family: Arial, sans-serif; padding: 20px;">
                    <h2>感谢注册考试系统</h2>
                    <p>请点击下方链接验证您的邮箱：</p>
                    <div style="padding: 15px; text-align: center;">
                        <a href="%s" style="display: inline-block; padding: 12px 32px; font-size: 16px; color: #fff; background: #1890ff; border-radius: 6px; text-decoration: none;">验证邮箱</a>
                    </div>
                    <p>如果按钮无法点击，请复制以下链接到浏览器：</p>
                    <p style="font-size: 12px; color: #666; word-break: break-all;">%s</p>
                    <p>如果您没有注册，请忽略此邮件。</p>
                </div>
                """.formatted(link, link);

        try {
            var message = mailSender.createMimeMessage();
            var helper = new MimeMessageHelper(message, true, "UTF-8");
            helper.setFrom(mailProperties.from());
            helper.setTo(to);
            helper.setSubject(subject);
            helper.setText(content, true);
            mailSender.send(message);
            log.info("验证邮件已发送至 {}", to);
        } catch (Exception e) {
            log.error("验证邮件发送失败: {}", to, e);
            throw new BusinessException("邮件发送失败");
        }
    }
}

File: exam-system-server/server/src/main/kotlin/com/yyjy/service/EmailService.kt
package com.yyjy.service

import com.yyjy.common.ExamProperties
import com.yyjy.constants.CacheConstant
import org.apache.commons.logging.LogFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.script.DefaultRedisScript
import org.springframework.http.HttpStatus
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessageHelper
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

@Service
class EmailService(
    private val mailSender: JavaMailSender,
    private val redisTemplate: RedisTemplate<String, Any>,
    private val examProperties: ExamProperties,
) {
    private val log = LogFactory.getLog(EmailService::class.java)
    /**
     * 发送注册验证码邮件
     * @param to 收件人邮箱
     * @param code 验证码
     */
    fun sendRegisterCaptcha(to: String, code: String) {
        val subject = "【考试系统】注册验证码"
        val content = """
            <div style="font-family: Arial, sans-serif; padding: 20px;">
                <h2>您好，欢迎注册考试系统</h2>
                <p>您的注册验证码为：</p>
                <div style="font-size: 24px; font-weight: bold; color: #1890ff; padding: 15px; background: #f0f9ff; border-radius: 8px; text-align: center;">
                    $code
                </div>
                <p>验证码有效期为 <strong>3 分钟</strong>，请勿泄露给他人。</p>
                <p>如果这不是您本人操作，请忽略此邮件。</p>
            </div>
        """.trimIndent()

        val message = mailSender.createMimeMessage()
        val helper = MimeMessageHelper(message, true, "UTF-8")
        helper.setFrom(examProperties.mail.from)
        helper.setTo(to)
        helper.setSubject(subject)
        helper.setText(content, true)
        mailSender.send(message)
    }

    /**
     * 生成验证码并发送
     * @param email 目标邮箱
     */
    fun generateAndSendCaptcha(email: String) {

        val key = "${CacheConstant.CAPTCHA_REGISTER_KEY}$email"

        // Lua 脚本：原子地检查 TTL 并写入
        // 返回 1 = 成功，0 = 被速率限制
        val luaScript = DefaultRedisScript(
            """
    local ttl = redis.call('TTL', KEYS[1])
    if ttl > 120 then
        return 0
    end
    redis.call('SET', KEYS[1], ARGV[1], 'EX', 180)
    return 1
    """.trimIndent(),
            Long::class.java
        )
        val code = (100000..999999).random().toString()
        val result = redisTemplate.execute(luaScript, listOf(key), code)
        if (result == 0L) {
            throw ResponseStatusException(
                HttpStatus.UNPROCESSABLE_ENTITY, "验证码已发送，请稍后再试"
            )
        }
        try {
            sendRegisterCaptcha(email, code)
        } catch (e: Exception) {
            redisTemplate.delete(key)
            log.error("验证码邮件发送失败: $email", e)
            throw ResponseStatusException(
                HttpStatus.INTERNAL_SERVER_ERROR, "邮件发送失败，请稍后重试"
            )
        }
        log.info("验证码已发送至 $email")
    }

    /**
     * 验证验证码是否正确
     * @param email 邮箱
     * @param code 用户输入的验证码
     * @return true if valid
     */
    fun verifyCaptcha(email: String, code: String): Boolean {
        val key = "${CacheConstant.CAPTCHA_REGISTER_KEY}$email"
        val storedCode = redisTemplate.opsForValue().get(key)
            ?: throw ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "验证码不存在")
        // 检查验证码是否匹配
        if (storedCode == code) {
            // 验证成功，清除验证码
            redisTemplate.delete(key)
            return true
        }
        throw ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "验证码错误")
    }
}